### PR TITLE
Don't crash when an interface is not implemented by the HostObject

### LINKED ISSUE #18633: NullReferenceException inside Microsoft.CodeAnalysis.BuildTasks.Csc.AddResponseFileCommands()
**Version Used**:
Microsoft.CodeAnalysis.* = 2.0.0
Visual Studio 2017
Inside .NET Framework 4.6.2 console application

**Steps to Reproduce**:

1. Install Microsoft.CodeAnalysis.CSharp.Workspaces into a .NET Framework 4.6.2 console application
2. Write the following code (the project being loaded here is a .NET Framework 4.6.2 library):
```
MSBuildWorkspace workspace = MSBuildWorkspace.Create();
workspace.OpenProjectAsync(@"E:\Code\Wyam\src\core\Wyam.Hosting\Wyam.Hosting.csproj").Result;
```

**Expected Behavior**:
The project loads into the workspace without problems.

**Actual Behavior**:
`workspace.Diagnostics` contains a diagnostic with the message `{[Failure] Msbuild failed when processing the file 'E:\Code\Wyam\src\core\Wyam.Hosting\Wyam.Hosting.csproj'}`

---

I first noticed this behavior in a more complex application that does run-time loading of project files into workspaces and logs any diagnostic messages. I was getting a ton of these `Msbuild failed when processing` messages, though they didn't seem to actually stop the project file from loading. I figured I'd attempt to reproduce in a simple application to make sure it wasn't something with my host (I do a lot of crazy assembly stuff, so I was concerned that was messing it up). I was able to repro:

2017-04-12_11h48_51

Exploring further, I turned on exception breaking at caught a `NullReferenceException` in `Microsoft.CodeAnalysis.BuildTasks.Csc.AddResponseFileCommands()`:

2017-04-12_10h54_43

A couple points:
* Notice that the `Microsoft.Build.Tasks.CodeAnalysis.dll` assembly is coming from my Visual Studio 2017 folder. Even if I install `Microsoft.Net.Compilers` (disabling the MSBuild tasks that actually use that package for compilation of the host application) and copy the `Microsoft.Build.Tasks.CodeAnalysis.dll` assembly out of the tools folder and into the bin, it still binds to the version in the Visual Studio 2017 path.
* The PDB and source code here is decompiled from the assembly in the Visual Studio 2017 path that I'm apparently linking to.
* I've highlighted the line I think is throwing the NRE.

My best guess (and I could be totally off) is that there's a type mismatch in here somewhere between the `ICscHostObject` interface we're trying to cast to and the actual one implemented by `this.HostObject` (which is a `CSharpCompilerInputs`). The cast fails and we get the NRE.

### PR DESCRIPTION
**Pull request by khyperia**
Partially resolves #18633, in that instead of crashing when the host object doesn't implement the `I{Vbc,Csc}HostObject` interface, we silently continue. This also fixes another error in the same file, but that fix is more concrete in that there was already logic for handling the error, it just wasn't getting hit (RCWForCurrentContext should throw a NullArgException in its constructor if passed null).

The decision of if we should silently eat the error or explicitly error out (for the first case) is up for debate. The only behavior from something correctly implementing from the interface is to suppress "/sqmsessionguid" during design time builds (if the HostObject does not implement the interface, it will not suppress the flag, which seemed to be the original intent of the code).

The issue still remains of how we got into this state in the first place (with HostObject not implementing `ICscHostObject` - possibly issues with assembly version mismatches), but this at least resolves the NRE.

Ping @agocke as they helped me work through this.

### CODE REVIEW

**Comment by agocke** [comment]
We probably want to retarget this PR to 15.6 as I don't think the bug would make it through escrow.

File: src/Compilers/Core/MSBuildTask/Csc.cs [hunk 1]
```diff
@@ -642,12 +641,12 @@ protected override HostObjectInitializationStatus InitializeHostObject()
...
                 // we can dynamically test for smarter interfaces later..
-                using (RCWForCurrentContext<ICscHostObject> hostObject = new RCWForCurrentContext<ICscHostObject>(HostObject as ICscHostObject))
+                if (HostObject is ICscHostObject hostObjectCOM)
                 {
-                    ICscHostObject cscHostObject = hostObject.RCW;
-
-                    if (cscHostObject != null)
+                    using (RCWForCurrentContext<ICscHostObject> hostObject = new RCWForCurrentContext<ICscHostObject>(hostObjectCOM))
                     {
+                        ICscHostObject cscHostObject = hostObject.RCW;
```
Comment (agocke) [inline src/Compilers/Core/MSBuildTask/Csc.cs]: Can this be null?

**Comment by agocke** [comment]
+ @jasonmalinowski As I believe the workspace is responsible for passing this possibly bogus host object. I'd also like your opinion on both the COM code change and the expectation of the workspace on what should come out of the compiler if the provided host object does not implement the proper interfaces.

**Comment by agocke** [comment]
@khyperia @jasonmalinowski Also, there should also probably be a follow-up bug on whomever owns workspaces as to why a bad host object is being constructed in the first place.

**Comment by jasonmalinowski** [comment]
I'm not sure why the compiler is making a code fix here at all. If you got passed something bad, it's the host's problem plain and simple.

I also need to wash my eyes out with soap now that I've seen RCWForCurrentContext. That looks...most suspicious.

**Comment by jasonmalinowski** [comment]
I guess @Pilchie why was this tagged in Compilers? Why wouldn't we just not pass the host object in whatever case this is in?

**Comment by Pilchie** [comment]
I don't really remember :/

File: src/Compilers/Core/MSBuildTask/Csc.cs (hunk 1 above)
Comment (khyperia) [inline src/Compilers/Core/MSBuildTask/Csc.cs]: Just checked, no - regardless of COM code, the getter for `RCW` throws if the underlying field is null. http://source.roslyn.io/#Microsoft.Build.Tasks.CodeAnalysis/RCWForCurrentContext.cs,81
(The constructor, where the COM stuff is, seems to also assert that it's not null after COM calls, using... the original object (???) if the calls fail)

**Comment by agocke** [comment]
@jasonmalinowski before, if the host object didn't implement ICscHostObject we would throw an NRE. We could change this to a class cast exception or something else or we could just pretend like no host object was given at all. The last option is what this pr implements. How would you feel about this behavior on the workspaces side?

**Comment by jasonmalinowski** [comment]
Ignoring for a moment that NullReferenceException isn't really the right exception to be throwing, I'd consider a throw if you can't handle the type to be morally equivalent to you throwing ArgumentException if somebody passed a bad call. If a host is setting a host object, and they're passing the wrong type, silently ignoring it doesn't seem to be terribly good.

I am further beginning to wonder if RCWForCurrentContext has bugs. 😦

**Comment by jasonmalinowski** [comment]
And I'd move this bug to IDE for further analysis; although yes a different exception might be better, it's still not your fault you're getting passed the wrong thing.

**Comment by jasonmalinowski** [comment]
I guess it's probably `InvalidOperationException` or something. It's not really an argument either.

**Comment by agocke** [comment]
@jasonmalinowski I don't know, HostObject is literally a property that you set. Seems like an argument to me.

**Comment by jasonmalinowski** [comment]
That's a property, not an argument. Don't make me quote compiler specs. :wink:

**Comment by agocke** [comment]
I think it's an argument now

**Comment by khyperia** [comment]
Ping @agocke for a quick re-review (added a commit that throws an exception instead of ignoring an incorrect type)

> We probably want to retarget this PR to 15.6 as I don't think the bug would make it through escrow.

What's the branch for this? There's no pleasantly-named branch with "15.5"/"15.6" in its name :)

**Comment by khyperia** [comment]
Ping @jasonmalinowski for a review

File: src/Compilers/Core/MSBuildTask/Csc.cs [hunk 2]
```diff
@@ -202,11 +202,14 @@ protected internal override void AddResponseFileCommands(CommandLineBuilderExten
...
             bool designTime = false;
-            if (HostObject != null)
+            if (HostObject is ICscHostObject csHost)
             {
-                var csHost = HostObject as ICscHostObject;
                 designTime = csHost.IsDesignTime();
             }
+            else if (HostObject != null)
+            {
+                throw new InvalidOperationException();
```
Comment (jaredpar) [inline src/Compilers/Core/MSBuildTask/Csc.cs]: Should we add a message here to indicate the reason?

Comment (khyperia) [inline src/Compilers/Core/MSBuildTask/Csc.cs]: Code style seems to be mixed here - when I was first implementing this, I think I managed to find a lot of examples of where `InvalidOperationException` had no message. However, doing a search over `Compilers/Core` (where this PR is) shows that (exactly!) half of `InvalidOperationException`s have arguments, and half don't.

Is there a reason for that, or would in a perfect world all of them have arguments? (i.e. should add a message here? If so, is there a style of messages I should follow? Localization concerns?)

Comment (jaredpar) [inline src/Compilers/Core/MSBuildTask/Csc.cs]: I think ideally these should all have messages with them. The only exception which doesn't require a message is items like UnreachableException. 

As for localization yeah they need to be localized. You can put the message in ErrorStrings.resx though, use the ID and everything else will be handled.

**Comment by khyperia** [comment]
retest windows_debug_vs-integration_prtest please

**Comment by khyperia** [comment]
Retest justification: I think xunit crashed on integration tests - `Roslyn Error: test timeout exceeded, dumping remaining processes`. Retesting to make sure it's flakiness. As a side note ( @jaredpar ?), I forget if we're actively looking for xunit crash dumps, but it looks like crash dumps succeeded to be written, at least according to jenkins logs.

**Comment by jaredpar** [comment]
Roslyn.VisualStudio.IntegrationTests.VisualBasic.BasicExpressionEvaluator.EvaluatePrimitiveValues

Again 😦 

Are you ready to merge? If so I'll override and merge given htat test is extremely flaky right now.

**Comment by dpoeschl** [comment]
retest windows_debug_vs-integration_prtest please

**Comment by khyperia** [comment]
> Are you ready to merge? If so I'll override and merge given that test is extremely flaky right now.

Yeah, I'm ready to merge. Looks like it failed *again* with `Roslyn Error: test timeout exceeded, dumping remaining processes`.

**Comment by dpoeschl** [comment]
retest windows_debug_vs-integration_prtest please

**Comment by dpoeschl** [comment]
I have a dump from the test hang and am investigating.

**Comment by dpoeschl** [comment]
But @jaredpar I wouldn't hesitate to force merge this. The failures are not out of the ordinary right now, unfortunately.

**Comment by jaredpar** [comment]
image

**Pull request merged**